sdk/ai/azure-ai-inference/src/main/java/com/azure/ai/inference/ChatCompletionClientTracer.java

```diff
@@ -0,0 +1,606 @@
+                        tracer.addEvent(eventName, eventAttributes, OffsetDateTime.now(ZoneOffset.UTC), span);
+                    }
+                }
+            }
+        }
+    }
+
+    private void traceCompletionResponseAttributes(ChatCompletions response, Context span) {
+        tracer.setAttribute("gen_ai.response.id", response.getId(), span);
+        tracer.setAttribute("gen_ai.response.model", response.getModel(), span);
+        final CompletionsUsage usage = response.getUsage();
+        if (usage != null) {
+            tracer.setAttribute("gen_ai.usage.input_tokens", usage.getPromptTokens(), span);
+            tracer.setAttribute("gen_ai.usage.output_tokens", usage.getCompletionTokens(), span);
+        }
+        final List<ChatChoice> choices = response.getChoices();
+        if (choices != null) {
+            tracer.setAttribute("gen_ai.response.finish_reasons", getFinishReasons(choices), span);
+        }
+    }
+
+    private void traceCompletionResponseAttributes(StreamingChatCompletionsUpdate response, String finishReasons,
+        Context span) {
+        tracer.setAttribute("gen_ai.response.id", response.getId(), span);
+        tracer.setAttribute("gen_ai.response.model", response.getModel(), span);
+        final CompletionsUsage usage = response.getUsage();
+        if (usage != null) {
+            tracer.setAttribute("gen_ai.usage.input_tokens", usage.getPromptTokens(), span);
+            tracer.setAttribute("gen_ai.usage.output_tokens", usage.getCompletionTokens(), span);
+        }
+        tracer.setAttribute("gen_ai.response.finish_reasons", finishReasons, span);
+    }
+
+    private void traceCompletionResponseEvents(ChatCompletions response, Context span) {
+        final List<ChatChoice> choices = response.getChoices();
+        if (choices != null) {
+            final OffsetDateTime now = OffsetDateTime.now(ZoneOffset.UTC);
+            for (ChatChoice choice : choices) {
+                traceChoiceEvent(toJsonString(choice), now, span);
+            }
+        }
+    }
+
+    private void traceChoiceEvent(String choiceContent, OffsetDateTime timestamp, Context span) {
+        final Map<String, Object> eventAttributes = new HashMap<>(2);
+        eventAttributes.put("gen_ai.system", INFERENCE_GEN_AI_SYSTEM_NAME);
+        eventAttributes.put("gen_ai.event.content", choiceContent);
+        tracer.addEvent("gen_ai.choice", eventAttributes, timestamp, span);
+    }
+
+    private String toJsonString(ChatRequestMessage message) {
+        try (ByteArrayOutputStream stream = new ByteArrayOutputStream();
+            JsonWriter writer = JsonProviders.createWriter(stream)) {
+            message.toJson(writer);
+            writer.flush();
+            return new String(stream.toByteArray(), StandardCharsets.UTF_8);
+        } catch (IOException e) {
+            logger.atWarning().log("'ChatRequestMessage' serialization error", e);
```

Comment:
let's log it with debug - there is nothing actionable for users. Or with warning but once for this tracer lifetime

Reply:
make sense, updated